[CI] RemoteClusterServiceTests.testCollectNodes fails on 5.x

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.x+multijob-unix-compatibility/os=debian/891/console

```
Suite: org.elasticsearch.transport.RemoteClusterServiceTests
  1> [2017-06-19T18:37:55,311][INFO ][o.e.t.RemoteClusterServiceTests] [testRemoteNodeAttribute]: before test
  1> [2017-06-19T18:37:55,336][INFO ][o.e.t.t.MockTransportService] [cluster_1_node_1] publish_address {127.0.0.1:9300}, bound_addresses {[::1]:9302}, {127.0.0.1:9300}
  1> [2017-06-19T18:37:55,346][INFO ][o.e.t.t.MockTransportService] [cluster_1_node_2] publish_address {127.0.0.1:9301}, bound_addresses {[::1]:9300}, {127.0.0.1:9301}
  1> [2017-06-19T18:37:55,355][INFO ][o.e.t.t.MockTransportService] [cluster_2_node_1] publish_address {127.0.0.1:9303}, bound_addresses {[::1]:9303}, {127.0.0.1:9303}
  1> [2017-06-19T18:37:55,360][INFO ][o.e.t.t.MockTransportService] [cluster_2_node_2] publish_address {127.0.0.1:9305}, bound_addresses {[::1]:9305}, {127.0.0.1:9305}
  1> [2017-06-19T18:37:55,365][INFO ][o.e.t.t.MockTransportService] publish_address {127.0.0.1:9306}, bound_addresses {[::1]:9307}, {127.0.0.1:9306}
  1> [2017-06-19T18:37:55,415][INFO ][o.e.t.RemoteClusterServiceTests] [testRemoteNodeAttribute]: after test
  1> [2017-06-19T18:37:55,423][INFO ][o.e.t.RemoteClusterServiceTests] [testRemoteClusterSeedSetting]: before test
  1> [2017-06-19T18:37:55,425][INFO ][o.e.t.RemoteClusterServiceTests] [testRemoteClusterSeedSetting]: after test
  1> [2017-06-19T18:37:55,431][INFO ][o.e.t.RemoteClusterServiceTests] [testGroupClusterIndices]: before test
  1> [2017-06-19T18:37:55,438][INFO ][o.e.t.t.MockTransportService] [cluster_1_node] publish_address {127.0.0.1:9303}, bound_addresses {[::1]:9303}, {127.0.0.1:9303}
  1> [2017-06-19T18:37:55,442][INFO ][o.e.t.t.MockTransportService] [cluster_2_node] publish_address {127.0.0.1:9304}, bound_addresses {[::1]:9304}, {127.0.0.1:9304}
  1> [2017-06-19T18:37:55,457][INFO ][o.e.t.t.MockTransportService] publish_address {127.0.0.1:9305}, bound_addresses {[::1]:9305}, {127.0.0.1:9305}
  1> [2017-06-19T18:37:55,512][INFO ][o.e.t.RemoteClusterServiceTests] [testGroupClusterIndices]: after test
  1> [2017-06-19T18:37:55,539][INFO ][o.e.t.RemoteClusterServiceTests] [testCollectNodes]: before test
  1> [2017-06-19T18:37:55,547][INFO ][o.e.t.t.MockTransportService] [cluster_1_node_1] publish_address {127.0.0.1:9302}, bound_addresses {[::1]:9302}, {127.0.0.1:9302}
  1> [2017-06-19T18:37:55,555][INFO ][o.e.t.t.MockTransportService] [cluster_1_node_2] publish_address {127.0.0.1:9303}, bound_addresses {[::1]:9303}, {127.0.0.1:9303}
  1> [2017-06-19T18:37:55,559][INFO ][o.e.t.t.MockTransportService] [cluster_2_node_1] publish_address {127.0.0.1:9304}, bound_addresses {[::1]:9301}, {127.0.0.1:9304}
  2> REPRODUCE WITH: gradle :core:test -Dtests.seed=2B5E6499090B67DF -Dtests.class=org.elasticsearch.transport.RemoteClusterServiceTests -Dtests.method="testCollectNodes" -Dtests.security.manager=true -Dtests.locale=de-GR -Dtests.timezone=Europe/Isle_of_Man
  1> [2017-06-19T18:37:55,563][INFO ][o.e.t.t.MockTransportService] [cluster_2_node_2] publish_address {127.0.0.1:9300}, bound_addresses {[::1]:9300}, {127.0.0.1:9300}
  1> [2017-06-19T18:37:55,566][INFO ][o.e.t.t.MockTransportService] publish_address {127.0.0.1:9301}, bound_addresses {[::1]:9304}, {127.0.0.1:9301}
  1> [2017-06-19T18:37:55,676][WARN ][o.e.t.RemoteClusterConnection] fetching nodes from external cluster cluster_2 failed
  1> org.elasticsearch.transport.RemoteTransportException: [][127.0.0.1:9300][cluster:monitor/state]
  1> Caused by: org.elasticsearch.transport.ActionNotFoundTransportException: No handler for action [cluster:monitor/state]
  1> 	at org.elasticsearch.transport.TcpTransport.handleRequest(TcpTransport.java:1471) ~[main/:?]
  1> 	at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:1369) ~[main/:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport.readMessage(MockTcpTransport.java:173) ~[framework-5.6.0-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport.access$1000(MockTcpTransport.java:70) ~[framework-5.6.0-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.lambda$doRun$0(MockTcpTransport.java:346) ~[framework-5.6.0-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.common.util.CancellableThreads.executeIO(CancellableThreads.java:105) ~[main/:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.doRun(MockTcpTransport.java:346) ~[framework-5.6.0-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_131]
  1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_131]
  1> 	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_131]
  1> [2017-06-19T18:37:55,694][WARN ][o.e.t.RemoteClusterConnection] fetching nodes from external cluster cluster_2 failed
  1> org.elasticsearch.transport.RemoteTransportException: [][127.0.0.1:9300][cluster:monitor/state]
  1> Caused by: org.elasticsearch.transport.ActionNotFoundTransportException: No handler for action [cluster:monitor/state]
  1> 	at org.elasticsearch.transport.TcpTransport.handleRequest(TcpTransport.java:1471) ~[main/:?]
  1> 	at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:1369) ~[main/:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport.readMessage(MockTcpTransport.java:173) ~[framework-5.6.0-SNAPSHOT.jar:?]
  2> NOTE: leaving temporary files on disk at: /var/lib/jenkins/workspace/elastic+elasticsearch+5.x+multijob-unix-compatibility/os/debian/core/build/testrun/test/J0/temp/org.elasticsearch.transport.RemoteClusterServiceTests_2B5E6499090B67DF-001
  1> 	at org.elasticsearch.transport.MockTcpTransport.access$1000(MockTcpTransport.java:70) ~[framework-5.6.0-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.lambda$doRun$0(MockTcpTransport.java:346) ~[framework-5.6.0-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.common.util.CancellableThreads.executeIO(CancellableThreads.java:105) ~[main/:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.doRun(MockTcpTransport.java:346) ~[framework-5.6.0-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_131]
  1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_131]
  2> NOTE: test params are: codec=Asserting(Lucene62): {}, docValues:{}, maxPointsInLeafNode=1672, maxMBSortInHeap=7.979932532818616, sim=RandomSimilarity(queryNorm=false,coord=no): {}, locale=de-GR, timezone=Europe/Isle_of_Man
  2> NOTE: Linux 3.16.0-4-amd64 amd64/Oracle Corporation 1.8.0_131 (64-bit)/cpus=4,threads=1,free=318609456,total=522715136
  1> 	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_131]
  1> [2017-06-19T18:37:55,715][INFO ][o.e.t.RemoteClusterServiceTests] [testCollectNodes]: after test
FAILURE 0.21s J0 | RemoteClusterServiceTests.testCollectNodes <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: class java.lang.IllegalStateException
   > 	at __randomizedtesting.SeedInfo.seed([2B5E6499090B67DF:CEDEF1023703CFD6]:0)
   > 	at org.elasticsearch.transport.RemoteClusterServiceTests.testCollectNodes(RemoteClusterServiceTests.java:451)
   > 	at java.lang.Thread.run(Thread.java:748)
  1> [2017-06-19T18:37:55,734][INFO ][o.e.t.RemoteClusterServiceTests] [testIncrementallyAddClusters]: before test
  1> [2017-06-19T18:37:55,739][INFO ][o.e.t.t.MockTransportService] [cluster_1_node] publish_address {127.0.0.1:9303}, bound_addresses {[::1]:9303}, {127.0.0.1:9303}
  1> [2017-06-19T18:37:55,755][INFO ][o.e.t.t.MockTransportService] [cluster_2_node] publish_address {127.0.0.1:9305}, bound_addresses {[::1]:9305}, {127.0.0.1:9305}
  1> [2017-06-19T18:37:55,763][INFO ][o.e.t.t.MockTransportService] publish_address {127.0.0.1:9307}, bound_addresses {[::1]:9307}, {127.0.0.1:9307}
  1> [2017-06-19T18:37:55,782][INFO ][o.e.t.RemoteClusterServiceTests] [testIncrementallyAddClusters]: after test
  1> [2017-06-19T18:37:55,785][INFO ][o.e.t.RemoteClusterServiceTests] [testPre53NodesAreRejected]: before test
  1> [2017-06-19T18:37:55,792][INFO ][o.e.t.t.MockTransportService] [cluster_1_node] publish_address {127.0.0.1:9303}, bound_addresses {[::1]:9303}, {127.0.0.1:9303}
  2> uestTests, DynamicMappingDisabledTests, StupidBackoffModelTests, TransportNodesActionTests, SuggestModeTests, MultiOrdinalsTests, FuzzyOptionsTests, SearchProfileShardResultsTests, TaskIdTests, CompoundAnalysisTests, CardinalityTests, DateFieldTypeTests, MatchPhrasePrefixQueryBuilderTests, MinHashFilterFactoryTests, InternalDerivativeTests, ScaledFloatFieldTypeTests, AutoCreateIndexTests, CreateIndexResponseTests, BoundTransportAddressTests, PathTrieTests, ShardFailedClusterStateTaskExecutorTests, InternalHDRPercentilesTests, CategoryQueryContextTests, ExtendedStatsBucketTests, BackoffPolicyTests, MoreLikeThisQueryBuilderTests, RecoveryDuringReplicationTests, NodeJoinControllerTests, FieldNamesFieldMapperTests, ClusterRerouteTests, InSyncAllocationIdTests, RestBuilderListenerTests, LongHashTests, SnowballAnalyzerTests, ShardUtilsTests, InternalGlobalTests, IndicesOptionsTests, ActionListenerTests, SearchHitTests, DeprecationLoggerTests, ReplicaAllocatedAfterPrimaryTests, SnapshotTests, NullValueObjectMappingTests, PlainHighlighterTests, RestAnalyzeActionTests, BinaryRangeAggregatorTests, DisableGraphQueryTests, SingleNodeDiscoveryTests, InputStreamIndexInputTests, PagedBytesReferenceTests, ClusterStateToStringTests, MetaDataStateFormatTests, UpdateResponseTests, IndicesLifecycleListenerSingleNodeTests, IndexNameExpressionResolverTests, InternalDateRangeTests, NodeInfoStreamingTests, IndexServiceTests, NodeEnvironmentTests, BinaryFieldMapperTests, BWCTemplateTests, ShardRoutingTests, SimpleAllTests, SettingTests, ScriptDocValuesDatesTests, SpanWithinQueryBuilderTests, XContentFuilderListenerTests, LongHashTests, SnowballAnalyzerTests, ShardUtilsTests, InternalGlobalTests, IndicesOptionsTests, ActionListenerTests, SearchHitTests, DeprecationLoggerTests, ReplicaAllocatedAfterPrimaryTests, SnapshotTests, NullValueObjectMappingTests, PlainHighlighterTests, RestAnalyzeActionTests, BinaryRangeAggregatorTests, DisableGraphQueryTests, SingleNodeDiscoveryTests, InputStreamIndexInputTests, PagedBytesReferenceTests, ClusterStateToStringTests, MetaDataStateFormatTests, UpdateResponseTests, IndicesLifecycleListenerSingleNodeTests, IndexNameExpressionResolverTests, InternalDateRangeTests, NodeInfoStreamingTests, IndexServiceTests, NodeEnvironmentTests, BinaryFieldMapperTests, BWCTemplateTests, ShardRoutingTests, SimpleAllTests, SettingTests, ScriptDocValuesDatesTests, SpanWithinQueryBuilderTests, XContentFactoryTests, InternalScriptedMetricTests, SortBuilderTests, ValueCountAggregatorTests, SearchPhaseControllerTests, BooleanFieldTypeTests, CommandTests, MultiPhrasePrefixQueryTests, NestedHelperTests, LegacyNumberFieldMapperTests, SearchOperationListenerTests, IndexRequestBuilderTests, BalanceUnbalancedClusterTests, PublishClusterStateActionTests, InternalDateHistogramTests, ReleasableBytesStreamOutputTests, CancellableThreadsTests, PluginsTests, XContentTypeTests, GeoDistanceRangeQueryTests, MurmurHash3Tests, ShardSearchFailureTests, SortedSetDVStringFieldDataTests, ThreadPoolStatsTests, PrimaryNotRelocatedWhileBeingRecoveredTests, SimulatePipelineRequestParsingTests, DynamicTemplatesTests, JavaVersionTests, MinAggregatorTests, IndexingOperationListenerTests, FailedNodeRoutingTests, CborXContentTests, InternalTopHitsTests, NumberFieldTypeTests, BoundedBreakIteratorScannerTests, ParentFieldTypeTests, CountedCollectorTests, URIPatternTests, QueryStringQueryBuilderTests, SettingsTests, StopTokenFilterTests, AsyncIOProcessorTests, SimulateProcessorResultTests, SearchRequestTests, RefCountedTests, JavaMultiFieldMergeTests, RangeFieldMapperTests, AllFieldMapperPositionIncrementGapTests, FileWatcherTests, PidFileTests, RestSearchScrollActionTests, ScrollContextTests, MetaDataTests, CopyOnWriteHashMapTests, SumBucketTests, CJKFilterFactoryTests, StoredScriptTests, InternalNestedTests, IndexSearcherWrapperTests, IngestMetadataTests, JvmInfoTests, DateTimeUnitTests, StringsTests, BlockingClusterStatePublishResponseHandlerTests, MultiLineStringBuilderTests, DirectoryUtilsTests, ClusterAllocationExplainActionTests, PipelineStoreTests, LiveVersionMapTests, TruncateTokenFilterTests, RoutingBackwardCompatibilityTests, BigArraysTests, PatternCaptureTokenFilterTests, DiskThresholdDeciderTests, ScriptSettingsTests, CumulativeSumTests, OperatorTests, TransportActionFilterChainTests, NetworkModuleTests, SearchResponseTests, NestedTests, GetStoredScriptRequestTests, StatsTests, GeoDistanceSortBuilderTests, AbstractSearchAsyncActionTests, UidFieldMapperTests, MultiFieldIncludeInAllMapperTests, KeepTypesFilterFactoryTests, TransportRolloverActionTests, CommonGramsTokenFilterFactoryTests, ScriptContextRegistryTests, TextFieldTypeTests, EnvelopeBuilderTests, StreamsTests, UpdateByQueryRequestTests, FuzzyQueryBuilderTests, MetaDataIndexAliasesServiceTests, PointBuilderTests, StoreTests, SearchSlowLogTests, LegacyByteFieldTypeTests, BulkRequestTests, CodecTests, ExplainRequestTests, PriorityComparatorTests, ClusterIndexHealthTests, FunctionScoreTests, ConstructingObjectParserTests, ClassPermissionTests, ValueCountTests, QueryRescoreBuilderTests, AllocationPriorityTests, IdFieldMapperTests, LegacyIntegerFieldTypeTests, MaxTests, DistanceUnitTests, ThrottlingAllocationTests, FieldMemoryStatsTests, BlobPathTests, MemoryCircuitBreakerTests, GeoDistanceTests, MessageDigestsTests, XMoreLikeThisTests, AddStringKeyStoreCommandTests, RemoveSettingKeyStoreCommandTests, ArrayUtilsTests, KeyedLockTests, PrioritizedRunnableTests, JacksonLocationTests, SimpleLuceneTests, ESFileStoreTests, GatewayMetaStateTests, MergePolicySettingsTests, CharMatcherTests, FieldDataCacheTests, FilterFieldDataTests, FieldStatsProviderRefreshTests, CopyToMapperTests, DocumentParserTests, IndexFieldMapperTests, IpFieldTypeTests, LegacyLongFieldTypeTests, LegacyStringMappingTests, UidTests, CommonTermsQueryParserTests, IndicesQueryBuilderTests, SpanNotQueryBuilderTests, SpanTermQueryBuilderTests, WildcardQueryBuilderTests, FunctionScoreQueryBuilderTests, ParentBulkByScrollTaskTests, DoubleNestedSortingTests, NestedSortingTests, FileInfoTests, IndicesClusterStateServiceRandomUpdatesTests, JvmGcMonitorServiceTests, InternalSettingsPreparerTests, BaseRestHandlerTests, RestUtilsTests, ScriptModesTests, AggregationsTests, IpRangeTests, SignificantTermsTests, FilterAggregatorTests, FilterTests, InternalMaxTests, InternalGeoBoundsTests, InternalSumTests, SearchSourceBuilderTests, SliceBuilderTests, SuggestionEntryTests, RegexOptionsTests, NoisyChannelSpellCheckerTests, RemoteClusterServiceT DocumentParserTests, IndexFieldMapperTests, IpFieldTypeTests, LegacyLongFieldTypeTests, LegacyStringMappingTests, UidTests, CommonTermsQueryParserTests, IndicesQueryBuilderTests, SpanNotQueryBuilderTests, SpanTermQueryBuilderTests, WildcardQueryBuilderTests, FunctionScoreQueryBuilderTests, ParentBulkByScrollTaskTests, DoubleNestedSortingTests, NestedSortingTests, FileInfoTests, IndicesClusterStateServiceRandomUpdatesTests, JvmGcMonitorServiceTests, InternalSettingsPreparerTests, BaseRestHandlerTests, RestUtilsTests, ScriptModesTests, AggregationsTests, IpRangeTests, SignificantTermsTests, FilterAggregatorTests, FilterTests, InternalMaxTests, InternalGeoBoundsTests, InternalSumTests, SearchSourceBuilderTests, SliceBuilderTests, SuggestionEntryTests, RegexOptionsTests, NoisyChannelSpellCheckerTests, RemoteClusterServiceTests]
  1> [2017-06-19T18:37:55,797][INFO ][o.e.t.t.MockTransportService] [cluster_2_node] publish_address {127.0.0.1:9304}, bound_addresses {[::1]:9304}, {127.0.0.1:9304}
  1> [2017-06-19T18:37:55,801][INFO ][o.e.t.t.MockTransportService] publish_address {127.0.0.1:9305}, bound_addresses {[::1]:9305}, {127.0.0.1:9305}
  1> [2017-06-19T18:37:55,849][INFO ][o.e.t.RemoteClusterServiceTests] [testPre53NodesAreRejected]: after test
  1> [2017-06-19T18:37:55,860][INFO ][o.e.t.RemoteClusterServiceTests] [testSettingsAreRegistered]: before test
  1> [2017-06-19T18:37:55,860][INFO ][o.e.t.RemoteClusterServiceTests] [testSettingsAreRegistered]: after test
  1> [2017-06-19T18:37:55,862][INFO ][o.e.t.RemoteClusterServiceTests] [testBuiltRemoteClustersSeeds]: before test
  1> [2017-06-19T18:37:55,864][INFO ][o.e.t.RemoteClusterServiceTests] [testBuiltRemoteClustersSeeds]: after test
Completed [990/992] on J0 in 0.58s, 8 tests, 1 failure <<< FAILURES!
```

I couldn't reproduce this locally, however:

gradle :core:test -Dtests.seed=2B5E6499090B67DF -Dtests.class=org.elasticsearch.transport.RemoteClusterServiceTests -Dtests.method="testCollectNodes" -Dtests.security.manager=true -Dtests.locale=de-GR -Dtests.timezone=Europe/Isle_of_Man

This time it failed differently then was initially recorded:

```
REPRODUCE WITH: gradle :core:test -Dtests.seed=7093F91C34F2CE93 -Dtests.class=org.elasticsearch.transport.RemoteClusterServiceTests -Dtests.method="testCollectNodes" -Dtests.security.manager=true -Dtests.jvm.argline="-XX:+AggressiveOpts" -Dtests.locale=en-CA -Dtests.timezone=US/Samoa
   > Throwable #1: java.lang.AssertionError
   > 	at org.elasticsearch.transport.RemoteClusterServiceTests.testCollectNodes(RemoteClusterServiceTests.java:412)
   > 	at java.lang.Thread.run(Thread.java:745)Throwable #2: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=2382, name=elasticsearch[__mock_network_thread][T#5], state=RUNNABLE, group=TGRP-RemoteClusterServiceTests]
   > Caused by: java.lang.AssertionError: should not be called
   > 	at __randomizedtesting.SeedInfo.seed([7093F91C34F2CE93]:0)
   > 	at org.elasticsearch.transport.RemoteClusterServiceTests$3.onResponse(RemoteClusterServiceTests.java:396)
   > 	at org.elasticsearch.transport.RemoteClusterServiceTests$3.onResponse(RemoteClusterServiceTests.java:392)
   > 	at org.elasticsearch.transport.RemoteClusterService$2.onResponse(RemoteClusterService.java:366)
   > 	at org.elasticsearch.transport.RemoteClusterService$2.onResponse(RemoteClusterService.java:359)
  2> NOTE: leaving temporary files on disk at: /var/lib/jenkins/workspace/elastic+elasticsearch+master+aggressive-opts/core/build/testrun/test/J1/temp/org.elasticsearch.transport.RemoteClusterServiceTests_7093F91C34F2CE93-001
  2> NOTE: test params are: codec=Asserting(Lucene70): {}, docValues:{}, maxPointsInLeafNode=633, maxMBSortInHeap=7.565153463230077, sim=RandomSimilarity(queryNorm=false): {}, locale=en-CA, timezone=US/Samoa
  2> NOTE: Linux 4.4.44-39.55.amzn1.x86_64 amd64/Oracle Corporation 1.8.0_121 (64-bit)/cpus=4,threads=1,free=397840256,total=519569408
   > 	at org.elasticsearch.transport.RemoteClusterConnection$2.handleResponse(RemoteClusterConnection.java:216)
   > 	at org.elasticsearch.transport.RemoteClusterConnection$2.handleResponse(RemoteClusterConnection.java:207)
   > 	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1053)
   > 	at org.elasticsearch.transport.TcpTransport$2.doRun(TcpTransport.java:1422)
   > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 	at org.elasticsearch.common.util.concurrent.EsExecutors$1.execute(EsExecutors.java:139)
   > 	at org.elasticsearch.transport.TcpTransport.handleResponse(TcpTransport.java:1414)
   > 	at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:1370)
   > 	at org.elasticsearch.transport.MockTcpTransport.readMessage(MockTcpTransport.java:171)
   > 	at org.elasticsearch.transport.MockTcpTransport.access$900(MockTcpTransport.java:72)
   > 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.lambda$doRun$0(MockTcpTransport.java:349)
   > 	at org.elasticsearch.common.util.CancellableThreads.executeIO(CancellableThreads.java:105)
   > 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.doRun(MockTcpTransport.java:349)
   > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
   > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
   > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
   > 	at java.lang.Thread.run(Thread.java:745)
```

Build url: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+aggressive-opts/318/consoleText

Another failure:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+java9-periodic/3057/console

This failure was a bit different: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-windows-compatibility/380/console

```
1> org.elasticsearch.transport.RemoteTransportException: [][127.0.0.1:52159][cluster:monitor/state]
  1> Caused by: org.elasticsearch.transport.ActionNotFoundTransportException: No handler for action [cluster:monitor/state]
  1> 	at org.elasticsearch.transport.TcpTransport.handleRequest(TcpTransport.java:1468) ~[main/:?]
  1> 	at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:1352) ~[main/:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport.readMessage(MockTcpTransport.java:171) ~[framework-6.0.0-alpha3-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport.access$900(MockTcpTransport.java:72) ~[framework-6.0.0-alpha3-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.lambda$doRun$0(MockTcpTransport.java:349) ~[framework-6.0.0-alpha3-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.common.util.CancellableThreads.executeIO(CancellableThreads.java:105) ~[main/:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.doRun(MockTcpTransport.java:349) ~[framework-6.0.0-alpha3-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_92]
  1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_92]
  1> 	at java.lang.Thread.run(Thread.java:745) ~[?:1.8.0_92]
  1> [2017-06-29T11:43:20,706][ERROR][o.e.t.RemoteClusterServiceTests] expected TransportException but got a different one see #25301

  1> java.lang.IllegalStateException: no seed node left
  1> 	at org.elasticsearch.transport.RemoteClusterConnection$ConnectHandler.collectRemoteNodes(RemoteClusterConnection.java:442) ~[main/:?]
  1> 	at org.elasticsearch.transport.RemoteClusterConnection$ConnectHandler$SniffClusterStateResponseHandler.handleException(RemoteClusterConnection.java:549) ~[main/:?]
  1> 	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1060) ~[main/:?]
  1> 	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1060) ~[main/:?]
  1> 	at org.elasticsearch.transport.TcpTransport.lambda$handleException$15(TcpTransport.java:1448) ~[main/:?]
  1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:569) ~[main/:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_92]
  1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_92]
  1> 	at java.lang.Thread.run(Thread.java:745) ~[?:1.8.0_92]
  1> [2017-06-29T11:43:21,682][INFO ][o.e.t.RemoteClusterServiceTests] [testCollectNodes]: after test

  2> REPRODUCE WITH: gradle :core:test -Dtests.seed=90380F809BA169EF -Dtests.class=org.elasticsearch.transport.RemoteClusterServiceTests -Dtests.method="testCollectNodes" -Dtests.security.manager=true -Dtests.locale=es-CO -Dtests.timezone=Australia/LHI
```

Another failure like https://github.com/elastic/elasticsearch/issues/25301#issuecomment-311901708 occurred: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-windows-compatibility/385/consoleFull

```
  2> WARNING: Uncaught exception in thread: Thread[elasticsearch[__mock_network_thread][T#7],5,TGRP-RemoteClusterServiceTests]
  2> java.lang.AssertionError: should not be called
  2> 	at __randomizedtesting.SeedInfo.seed([DE3388ADB60D645B]:0)
  2> 	at org.junit.Assert.fail(Assert.java:88)
  2> 	at org.elasticsearch.transport.RemoteClusterServiceTests$3.onResponse(RemoteClusterServiceTests.java:396)
  2> 	at org.elasticsearch.transport.RemoteClusterServiceTests$3.onResponse(RemoteClusterServiceTests.java:392)
  2> 	at org.elasticsearch.transport.RemoteClusterService$2.onResponse(RemoteClusterService.java:366)
  2> 	at org.elasticsearch.transport.RemoteClusterService$2.onResponse(RemoteClusterService.java:359)
  2> 	at org.elasticsearch.transport.RemoteClusterConnection$2.handleResponse(RemoteClusterConnection.java:216)
  2> 	at org.elasticsearch.transport.RemoteClusterConnection$2.handleResponse(RemoteClusterConnection.java:207)
  2> 	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1053)
  2> 	at org.elasticsearch.transport.TcpTransport$2.doRun(TcpTransport.java:1422)
  2> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
  2> 	at org.elasticsearch.common.util.concurrent.EsExecutors$1.execute(EsExecutors.java:139)
  2> 	at org.elasticsearch.transport.TcpTransport.handleResponse(TcpTransport.java:1414)
  2> 	at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:1370)
  2> 	at org.elasticsearch.transport.MockTcpTransport.readMessage(MockTcpTransport.java:171)
  2> 	at org.elasticsearch.transport.MockTcpTransport.access$900(MockTcpTransport.java:72)
  2> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.lambda$doRun$0(MockTcpTransport.java:349)
  2> 	at org.elasticsearch.common.util.CancellableThreads.executeIO(CancellableThreads.java:105)
  2> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.doRun(MockTcpTransport.java:349)
  2> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
  2> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  2> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  2> 	at java.lang.Thread.run(Thread.java:745)
```

This does not reproduce on my laptop:

```
gradle :core:test -Dtests.seed=DE3388ADB60D645B -Dtests.class=org.elasticsearch.transport.RemoteClusterServiceTests -Dtests.method="testCollectNodes" -Dtests.security.manager=true -Dtests.locale=sq -Dtests.timezone=Etc/GMT+10
```

And the other type of failure (like https://github.com/elastic/elasticsearch/issues/25301#issuecomment-311919795) occurs here: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-windows-compatibility/386/consoleFull

```
  1> org.elasticsearch.transport.RemoteTransportException: [][127.0.0.1:57632][cluster:monitor/state]
  1> Caused by: org.elasticsearch.transport.ActionNotFoundTransportException: No handler for action [cluster:monitor/state]
  1> 	at org.elasticsearch.transport.TcpTransport.handleRequest(TcpTransport.java:1468) ~[main/:?]
  1> 	at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:1352) ~[main/:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport.readMessage(MockTcpTransport.java:171) ~[framework-6.0.0-alpha3-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport.access$900(MockTcpTransport.java:72) ~[framework-6.0.0-alpha3-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.lambda$doRun$0(MockTcpTransport.java:349) ~[framework-6.0.0-alpha3-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.common.util.CancellableThreads.executeIO(CancellableThreads.java:105) ~[main/:?]
  1> 	at org.elasticsearch.transport.MockTcpTransport$MockChannel$1.doRun(MockTcpTransport.java:349) ~[framework-6.0.0-alpha3-SNAPSHOT.jar:?]
  1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[main/:?]
  1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_121]
  1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_121]
  1> 	at java.lang.Thread.run(Thread.java:745) ~[?:1.8.0_121]
```

This also does not reproduce on my laptop:

```
gradle :core:test -Dtests.seed=1FBC3607DA4E1B19 -Dtests.class=org.elasticsearch.transport.RemoteClusterServiceTests -Dtests.method="testCollectNodes" -Dtests.security.manager=true -Dtests.locale=pt -Dtests.timezone=America/Bogota
```

Another failure?
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.x+multijob-unix-compatibility/os=amazon/934/console

	2> WARNING: Uncaught exception in thread: Thread[elasticsearch[__mock_network_thread][T#7],5,TGRP-RemoteClusterServiceTests]
	  2> java.lang.AssertionError: should not be called
	  2> 	at __randomizedtesting.SeedInfo.seed([DF0D11B11B257AB3]:0)
	  1> [2017-07-03T06:09:21,358][INFO ][o.e.t.t.MockTransportService] [cluster_2_node_1] publish_address {127.0.0.1:9303}, bound_addresses {[::1]:9304}, {127.0.0.1:9303}
	  1> [2017-07-03T06:09:21,363][INFO ][o.e.t.t.MockTransportService] [cluster_2_node_2] publish_address {127.0.0.1:9305}, bound_addresses {[::1]:9305}, {127.0.0.1:9305}
	  1> [2017-07-03T06:09:21,370][INFO ][o.e.t.t.MockTransportService] publish_address {127.0.0.1:9306}, bound_addresses {[::1]:9307}, {127.0.0.1:9306}
	  2> 	at org.junit.Assert.fail(Assert.java:88)
	  2> 	at org.elasticsearch.transport.RemoteClusterServiceTests$3.onResponse(RemoteClusterServiceTests.java:434)
	  2> 	at org.elasticsearch.transport.RemoteClusterServiceTests$3.onResponse(RemoteClusterServiceTests.java:430)
	  2> 	at org.elasticsearch.transport.RemoteClusterService$2.onResponse(RemoteClusterService.java:377)
	  1> [2017-07-03T06:09:21,493][INFO ][o.e.t.RemoteClusterServiceTests] [testRemoteNodeAttribute]: after test
	  2> 	at org.elasticsearch.transport.RemoteClusterService$2.onResponse(RemoteClusterService.java:370)
	  2> 	at org.elasticsearch.transport.RemoteClusterConnection$2.handleResponse(RemoteClusterConnection.java:216)
	  1> [2017-07-03T06:09:21,499][INFO ][o.e.t.RemoteClusterServiceTests] [testRemoteClusterSeedSetting]: before test
	  2> 	at org.elasticsearch.transport.RemoteClusterConnection$2.handleResponse(RemoteClusterConnection.java:207)

This did not reproduce on my laptop:

    gradle :core:test -Dtests.seed=DF0D11B11B257AB3 -Dtests.class=org.elasticsearch.transport.RemoteClusterServiceTests -Dtests.method="testCollectNodes" -Dtests.security.manager=true -Dtests.locale=ko -Dtests.timezone=US/Eastern

--- a/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java
@@ -19,6 +19,7 @@
 
 package org.elasticsearch.test.transport;
 
+import com.carrotsearch.randomizedtesting.SysGlobals;
 import org.elasticsearch.Version;
 import org.elasticsearch.cluster.ClusterModule;
 import org.elasticsearch.cluster.node.DiscoveryNode;
@@ -43,6 +44,7 @@ import org.elasticsearch.indices.breaker.NoneCircuitBreakerService;
 import org.elasticsearch.node.Node;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.tasks.TaskManager;
+import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.test.tasks.MockTaskManager;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.ConnectTransportException;
@@ -89,6 +91,7 @@ import java.util.function.Function;
 public final class MockTransportService extends TransportService {
 
     private final Map<DiscoveryNode, List<Transport.Connection>> openConnections = new HashMap<>();
+    private static final int JVM_ORDINAL = Integer.parseInt(System.getProperty(SysGlobals.CHILDVM_SYSPROP_JVM_ID, "0"));
 
     public static class TestPlugin extends Plugin {
         @Override
@@ -99,6 +102,12 @@ public final class MockTransportService extends TransportService {
 
     public static MockTransportService createNewService(Settings settings, Version version, ThreadPool threadPool,
             @Nullable ClusterSettings clusterSettings) {
+        // some tests use MockTransportService to do network based testing. Yet, we run tests in multiple JVMs that means
+        // concurrent tests could claim port that another JVM just released and if that test tries to simulate a disconnect it might
+        // be smart enough to re-connect depending on what is tested. To reduce the risk, since this is very hard to debug we use
+        // a different default port range per JVM unless the incoming settings override it
+        int basePort = 9300 + (JVM_ORDINAL * 100);
+        settings = Settings.builder().put(TcpTransport.PORT.getKey(), basePort + "-" + (basePort+100)).put(settings).build();
         NamedWriteableRegistry namedWriteableRegistry = new NamedWriteableRegistry(ClusterModule.getNamedWriteables());
         final Transport transport = new MockTcpTransport(settings, threadPool, BigArrays.NON_RECYCLING_INSTANCE,
                 new NoneCircuitBreakerService(), namedWriteableRegistry, new NetworkService(Collections.emptyList()), version);

